src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,240 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
+            ILogger log,
+            CancellationToken token)
+        {
+            var metadataCache = new MetadataReferenceCache();
+            var registrationUri = _regResource.GetUri(packageId);
+
+            var registrationIndexResult = await LoadRangesAsync(
+                _client,
+                registrationUri,
+                packageId,
+                sourceCacheContext,
+                async httpSourceResult => await ProcessRegistrationIndexAsync<RegistrationIndex>(httpSourceResult, token),
+                log,
+                token);
+
+            var registrationIndex = registrationIndexResult.Index;
+            var httpSourceCacheContext = registrationIndexResult.CacheContext;
+
+            if (registrationIndex == null)
+            {
+                // The server returned a 404, the package does not exist
+                return Enumerable.Empty<PackageSearchMetadataRegistration>();
+            }
+
+            var results = new List<PackageSearchMetadataRegistration>();
+
+            foreach (var registrationPage in registrationIndex.Items)
+            {
+                if(registrationPage==null)
+                {
+                    throw new InvalidDataException(registrationUri.AbsoluteUri);
+                }
+
+                var lower = NuGetVersion.Parse(registrationPage.Lower);
+                var upper = NuGetVersion.Parse(registrationPage.Upper);
+
+                if (range.IsItemRangeRequired(lower, upper))
+                {
+                    if (registrationPage.Items == null)
+                    {
+                        var rangeUri = registrationPage.Url;
+                        var leafRegistrationPage = await GetRegistratioIndexPageAsync(_client, rangeUri, packageId, lower, upper, httpSourceCacheContext, log, token);
+
+                        if(registrationPage==null)
+                        {
+                            throw new InvalidDataException(registrationUri.AbsoluteUri);
+                        }
+
+                        ProcessRegistrationPage(leafRegistrationPage, results, range, includePrerelease,includeUnlisted, metadataCache);
+                    }
+                    else
+                    {
+                        ProcessRegistrationPage(registrationPage, results, range, includePrerelease, includeUnlisted, metadataCache);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Process RegistrationIndex and return list of RegistrationPages.
+        /// </summary>
+        /// <typeparam name="T">Generic type</typeparam>
+        /// <param name="httpSourceResult">HttpSourceResult type object which comes from HttpSource, used for get stream.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        internal async Task<T> ProcessRegistrationIndexAsync<T>(HttpSourceResult httpSourceResult, CancellationToken token)
```

Comment:
Is there any point in accepting a token if it's not used? Perhaps you could throw if cancelled on the first line to avoid creating readers/serializes if cancellation already occurred

Reply:
Added `token.ThrowIfCancellationRequested();`